### DISCUSSION TITLE
Porous media twophase.i

### DISCUSSION THREAD

**Discussion by POLonge** [idaholab/moose#26652] [Q&A Modules: Porous Flow]
Hi,

I'm having this error with my codes and MSG back. What does this mean?

*** ERROR ***
A 'PorousFlowDictator' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

@GiudGiud @cpgr

**Comment by 1runer** [comment]
Did you compile your own application with the PorousFlow module activated in the Makefile?
You should have an entry like this:

POROUS_FLOW                 := yes

**Reply by POLonge** [reply]
Yeah, I did that on the Makefile

################################## MODULES ####################################
ALL_MODULES                 := no

CHEMICAL_REACTIONS          := no
CONTACT                     := no
ELECTROMAGNETICS            := no
EXTERNAL_PETSC_SOLVER       := no
FLUID_PROPERTIES            := no
FSI                         := no
FUNCTIONAL_EXPANSION_TOOLS  := no
GEOCHEMISTRY                := no
HEAT_TRANSFER               := no
LEVEL_SET                   := no
MISC                        := no
NAVIER_STOKES               := yes
OPTIMIZATION                := no
PERIDYNAMICS                := no
PHASE_FIELD                 := no
POROUS_FLOW                 := yes
RAY_TRACING                 := no
REACTOR                     := no
RDG                         := no
RICHARDS                    := no
STOCHASTIC_TOOLS            := no
THERMAL_HYDRAULICS          := no
TENSOR_MECHANICS            := no
XFEM                        := no

**Comment by 1runer** [comment]
The message tells you that your app doesn't find the PorousFlowDictator which should be an object in the input-file of your simulation

E.g. like in this [exemplary file](https://github.com/idaholab/moose/blob/796cb8aa68a15d61eb82eb353bb6582b01c6634f/modules/porous_flow/test/tests/jacobian/brineco2_twophase.i)
```
[UserObjects]
  [dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pgas zi'
    number_fluid_phases = 2
    number_fluid_components = 2
  []
[]
```

Do the example or test files run?

**Reply by POLonge** [reply]
This is what I have in my code
```
[UserObjects]
      [./dictator]
            type = PorousFlowDictator
            porous_flow_vars = 'pwater pgas T disp_x disp_y'
            number_fluid_phases = 2
            number_fluid_components = 2
          [../]
          [./pc]
            type = PorousFlowCapillaryPressureVG
            alpha = 2.66E-6
            m = 0.6
  [../]
[]
```

**Reply by GiudGiud** [reply]
Once you enable the module in the makefile you should make clean then recompile. 

The new executable created will have access to that object

**Comment by POLonge** [comment]
Hi @GiudGiud @1runer 

I'm having similar error with the BC

*** ERROR ***
A 'PresetBC' is not a registered object.

If you are trying to find this object in a dynamically loaded library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by GiudGiud** [reply]
hello

where did you find this name for this BC? It's not offered in MOOSE

Guillaume

**Reply by POLonge** [reply]
Yeah I got that from an old existing code. This is resolved.

**Comment by POLonge** [comment]
Hi @GiudGiud @1runer
I'm having this error, how do i fix it. Code is below
```
*** ERROR ***
/home/p815l321/projects/sim/problems/twophase.i:548.1: syntax error, unexpected end of file

```

```
# Test of 2Phase Flow
[Mesh]
 [./annular]
   type = AnnularMeshGenerator
   dim = 2
   nr = 25
   rmin = 1
   rmax = 10
   nt = 25
   tmin =0
   tmax =3.12
   growth_r = 1.3
  [../]

  [./res]
    type = SubdomainBoundingBoxGenerator
    input = annular
    block_id = 0
    
  [./vfrac1]
    type = SubdomainBoundingBoxGenerator
    input = annular
    block_id = 1
    bottom_left = '5 3 -5'
    top_right = '5.5 3.5 5'
    
  [../]


  [./vfrac2]
   type = SubdomainBoundingBoxGenerator
   input = annular
   block_id = 2
   bottom_left = '-5.5 3 -5'
   top_right = '-5 3.5 5'
  [../]


  [./vug1]
   type = SubdomainBoundingBoxGenerator
   input = annular
   block_id = 3
   bottom_left = '1 1 -1'
   top_right = '1.5 1.5 0'
  [../]

  [./vug2]
   type = SubdomainBoundingBoxGenerator
   input = annular
   block_id = 4
   bottom_left = '-3.5 0 -1'
   top_right = '-3 1 0'
   
  [../]

  [./vug3]
   type = SubdomainBoundingBoxGenerator
   input = annular
   block_id = 5
   bottom_left = '0 4 0'
   top_right = '0.5 4.5 .5'
   
  [../]

  [./rename]
    type = RenameBlockGenerator
    input = annular
    old_block_id = '0 1 2 3 4 5'
    new_block_name = 'res vfrac1 vfrac2 vug1 vug2 vug3'
  [../]
[]

[UserObjects]
  [./dictator]
    type = PorousFlowDictator
    porous_flow_vars = 'pwater pgas T disp_x disp_y'
    number_fluid_phases = 2
    number_fluid_components = 2
  [../]
  [./pc]
    type = PorousFlowCapillaryPressureVG
    alpha = 2.66E-6
    m = 0.6
  [../]
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  gravity = '0 0 0'
  biot_coefficient = 1
  PorousFlowDictator = dictator
[]

[Variables]
  [./pwater]
    initial_condition = 20E6
  [../]
  [./pgas]
    initial_condition = 20.5E6
  [../]
  [./T]
    initial_condition = 330
    scaling = 1E-5
  [../]
  [./disp_x]
    scaling = 1E-5
  [../]
  [./disp_y]
    scaling = 1E-5
  [../]
[]

[Kernels]
  [./mass_water_dot]
    type = PorousFlowMassTimeDerivative
    fluid_component = 0
    variable = pwater
  [../]
  [./flux_water]
    type = PorousFlowAdvectiveFlux
    fluid_component = 0
    variable = pwater
  [../]
  [./vol_strain_rate_water]
    type = PorousFlowMassVolumetricExpansion
    fluid_component = 0
    variable = pwater
  [../]
  [./mass_co2_dot]
    type = PorousFlowMassTimeDerivative
    fluid_component = 1
    variable = pgas
  [../]
  [./flux_co2]
    type = PorousFlowAdvectiveFlux
    fluid_component = 1
    variable = pgas
  [../]
  [./vol_strain_rate_co2]
    type = PorousFlowMassVolumetricExpansion
    fluid_component = 1
    variable = pgas
  [../]
  [./energy_dot]
    type = PorousFlowEnergyTimeDerivative
    variable = T
  [../]
  [./advection]
    type = PorousFlowHeatAdvection
    variable = T
  [../]
  [./conduction]
    type = PorousFlowHeatConduction
    variable = T
  [../]
  [./vol_strain_rate_heat]
    type = PorousFlowHeatVolumetricExpansion
    variable = T
  [../]
  [./grad_stress_x]
    type = StressDivergenceTensors
    temperature = T
    variable = disp_x
    thermal_eigenstrain_name = thermal_contribution
    component = 0
  [../]
  [./poro_x]
    type = PorousFlowEffectiveStressCoupling
    variable = disp_x
    component = 0
  [../]
  [./grad_stress_y]
    type = StressDivergenceTensors
    temperature = T
    variable = disp_y
    thermal_eigenstrain_name = thermal_contribution
    component = 1
  [../]
  [./poro_y]
    type = PorousFlowEffectiveStressCoupling
    variable = disp_y
    component = 1
  [../]
[]

[AuxVariables]
  [./disp_z]
  [../]
  [./effective_fluid_pressure]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./mass_frac_phase0_species0]
    initial_condition = 1 # all water in phase=0
  [../]
  [./mass_frac_phase1_species0]
    initial_condition = 0 # no water in phase=1
  [../]
  [./sgas]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = 0.7
  [../]
  [./swater]
    family = MONOMIAL
    order = CONSTANT
    initial_condition = 0.3
  [../]
  [./stress_rr]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./stress_tt]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./stress_zz]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./porosity]
    family = MONOMIAL
    order = CONSTANT
  [../]
[]

[AuxKernels]
  [./effective_fluid_pressure]
    type = ParsedAux
    args = 'pwater pgas swater sgas'
    function = 'pwater * swater + pgas * sgas'
    variable = effective_fluid_pressure
  [../]
  
  [./swater]
    type = PorousFlowPropertyAux
    variable = swater
    property = saturation
    phase = 0
    execute_on = timestep_end
  [../]
  
  [./sgas]
    type = PorousFlowPropertyAux
    variable = sgas
    property = saturation
    phase = 1
    execute_on = timestep_end
  [../]
  
  [./stress_rr]
    type = RankTwoScalarAux
    variable = stress_rr
    rank_two_tensor = stress
    scalar_type = RadialStress
    point1 = '0 0 0'
    point2 = '0 0 1'
    execute_on = timestep_end
  [../]
  
  [./stress_tt]
    type = RankTwoScalarAux
    variable = stress_tt
    rank_two_tensor = stress
    scalar_type = HoopStress
    point1 = '0 0 0'
    point2 = '0 0 1'
    execute_on = timestep_end
  [../]
  
  [./stress_zz]
    type = RankTwoAux
    variable = stress_zz
    rank_two_tensor = stress
    index_i = 2
    index_j = 2
    execute_on = timestep_end
  [../]
  
  [./porosity]
    type = MaterialRealAux
    variable = porosity
    property = PorousFlow_porosity_qp
    execute_on = timestep_end
  [../]
[]


[BCs]
  [./roller_tmax]
    type = FunctionDirichletBC
    variable = disp_x
    value = 0
    boundary = top
  [../]

  [./roller_tmin]
    type = FunctionDirichletBC
    variable = disp_y
    value = 0
    boundary = bottom
  [../]

  [./pinned_top_bottom_x]
    type = FunctionDirichletBC
    variable = disp_x
    value = 0
    boundary = 'top bottom'
  [../]

  [./pinned_top_bottom_y]
    type = FunctionDirichletBC
    variable = disp_y
    value = 0
    boundary = 'top bottom'
  [../]
  
  [./cavity_pressure_x]
    type = Pressure
    boundary = top
    variable = disp_x
    component = 0
    postprocessor = constrained_effective_fluid_pressure_at_wellbore 
  [../]
  
  [./cavity_pressure_y]
    type = Pressure
    boundary = top
    variable = disp_y
    component = 1
    postprocessor = constrained_effective_fluid_pressure_at_wellbore
  [../]

  [./cold_co2]
    type = FunctionDirichletBC
    boundary = 'rmin'
    variable = T
    value = 290  # injection temperature
  [../]

  [./constant_co2_inj]
    type = PorousFlowSink
    boundary = rmin
    variable = pgas
    fluid_phase = 1
    flux_function = -5E-4
  [../]

[]

[Modules]
  [./FluidProperties]
    [./true_water]
      type = Water97FluidProperties
    [../]
    
    [./tabulated_water]
      type = TabulatedBicubicFluidProperties
      fp = true_water
      temperature_min = 275
      pressure_max = 1E8
      interpolated_properties = 'density viscosity enthalpy internal_energy'
      fluid_property_file = water97_tabulated_11.csv
    [../]
    
    [./true_co2]
      type = CO2FluidProperties
    [../]
    
    [./tabulated_co2]
      type = TabulatedBicubicFluidProperties
      fp = true_co2
      temperature_min = 275
      pressure_max = 1E8
      interpolated_properties = 'density viscosity enthalpy internal_energy'
      fluid_property_file = co2_tabulated_11.csv
    [../]
    
  [../]
[]

[Materials]
  [./temperature]
    type = PorousFlowTemperature
    temperature = T
  [../]
  
  [./saturation_calculator]
    type = PorousFlow2PhasePP
    phase0_porepressure = pwater
    phase1_porepressure = pgas
    capillary_pressure = pc
  [../]
  
  [./massfrac]
    type = PorousFlowMassFraction
    mass_fraction_vars = 'mass_frac_phase0_species0 mass_frac_phase1_species0'
  [../]
  
  [./water]
    type = PorousFlowSingleComponentFluid
    fp = tabulated_water
    phase = 0
  [../]
  
  [./co2]
    type = PorousFlowSingleComponentFluid
    fp = tabulated_co2
    phase = 1
  [../]
  
  [./relperm_water]
    type = PorousFlowRelativePermeabilityCorey
    n = 4
    s_res = 0.1
    sum_s_res = 0.2
    phase = 0
  [../]
  
  [./relperm_co2]
    type = PorousFlowRelativePermeabilityBC
    nw_phase = true
    lambda = 2
    s_res = 0.1
    sum_s_res = 0.2
    phase = 1
 [../]

  [./porosity]
    type = PorousFlowPorosity
    fluid = true
    mechanical = true
    thermal = true
    porosity_zero = 0.1
    reference_temperature = 330
    reference_porepressure = 20E6
    thermal_expansion_coeff = 15E-6 # volumetric
    solid_bulk = 8E9 # unimportant since biot = 1
    block = res
  [../]

  [./porosityopen]
   type = PorousFlowPorosity
   porosity_zero = 0.9999
   block = 'vfrac1 vfrac2 vug1 vug2 vug3'
  [../]


  [./permeability_open]
    type = PorousFlowPermeabilityKozenyCarman
    block = 'vfrac1 vfrac2 vug1 vug2 vug3'
    poroperm_function = kozeny_carman_phi0
    phi0 = 0.9999
    n = 2
    m = 2
    k0 = 1E-15
  [../]

  [./permeability_res]
    type = PorousFlowPermeabilityConst
    block = res
    permeability = '1E-15 0 0  0 1E-15 0  0 0 1E-16'
  [../]

  [./rock_thermal_conductivity]
    type = PorousFlowThermalConductivityIdeal
    dry_thermal_conductivity = '2 0 0  0 2 0  0 0 2'
  [../]
  
  [./rock_internal_energy]
    type = PorousFlowMatrixInternalEnergy
    specific_heat_capacity = 1100
    density = 2300
  [../]

  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 66E9
    poissons_ratio = 0.18
  [../]
  
  [./strain]
    type = ComputeSmallStrain
    eigenstrain_names = 'thermal_contribution initial_stress'
  [../]
  
  [./thermal_contribution]
    type = ComputeThermalExpansionEigenstrain
    temperature = T
    thermal_expansion_coeff = 5E-6 # this is the linear thermal expansion coefficient
    eigenstrain_name = thermal_contribution
    stress_free_temperature = 330
  [../]
  
  [./initial_strain]
    type = ComputeEigenstrainFromInitialStress
    initial_stress = '20E6 0 0  0 20E6 0  0 0 20E6'
    eigenstrain_name = initial_stress
  [../]
  
  [./stress]
    type = ComputeLinearElasticStress
  [../]

  [./effective_fluid_pressure]
    type = PorousFlowEffectiveFluidPressure
  [../]
  
  [./volumetric_strain]
    type = PorousFlowVolumetricStrain
  [../]
  
[]

[Postprocessors]
  [./effective_fluid_pressure_at_wellbore]
    type = PointValue
    variable = effective_fluid_pressure
    point = '1 0 0'
    execute_on = timestep_begin
    use_displaced_mesh = false
  [../]
  
  [./constrained_effective_fluid_pressure_at_wellbore]
    type = FunctionValuePostprocessor
    function = constrain_effective_fluid_pressure
    execute_on = timestep_begin
  [../]
[]

[Functions]
  [./constrain_effective_fluid_pressure]
    type = ParsedFunction
    vars = effective_fluid_pressure_at_wellbore
    vals = effective_fluid_pressure_at_wellbore
    value = 'max(effective_fluid_pressure_at_wellbore, 20E6)'
  [../]
[]

[Preconditioning]
  active = basic
  [./basic]
    type = SMP
    full = true
    petsc_options = '-ksp_diagonal_scale -ksp_diagonal_scale_fix'
    petsc_options_iname = '-pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap'
    petsc_options_value = ' asm      lu           NONZERO                   2'
  [../]
  [./preferred_but_might_not_be_installed]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = ' lu       mumps'
  [../]
[]

[Executioner]
  type = Transient
  solve_type = Newton
  end_time = 2.6E6
  dt = 250
  nl_abs_tol = 1E-10
  nl_max_its = 10
  l_max_its = 25
[]

[Outputs]
 exodus = true
 file_base = twophaseresults
[]
```

**Reply by 1runer** [reply]
The error message is pretty simple.
`*** ERROR ***
/home/p815l321/projects/sim/problems/twophase.i:548.1: syntax error, unexpected end of file`

It tells you that you have syntax error in line 548 of your original input file. So there is something MOOSE doesn't recognize as MOOSE language.

Btw. if this file is pretty old (it looks like 2020 or older), there are good chances that you have to modify much more of your syntax to be consistent with recent MOOSE.

**Reply by POLonge** [reply]
Hi @GiudGiud @1runer 
Thanks so much for previous clarification.

This time, I'm stuck with error like these when trying to create fractures and cavity in my mesh, with this codes
  [vfrac1]
    type = SubdomainBoundingBox
    block_id = 4
    top_right = '6.5 6.5 9'
    bottom_left = '6 6 1'
  []
  [nodevfrac1]
    type = BoundingBoxNodeSet
    new_boundary = 'nodesvfrac1'
    bottom_left = '6 6 1'
    top_right = '6.5 6.5 9'
  []
  [vfrac2]
    type = SubdomainBoundingBox
    block_id = 5
    top_right = '2.5 2.5 6 '
    bottom_left = '2 2 1'
  []
  [nodevfrac2]
    type = BoundingBoxNodeSet
    new_boundary = 'nodesvfrac2'
    bottom_left = '2 2 1'
    top_right = '2.5 2.5 6'
  []
  [vug1]
    type = SubdomainBoundingBox
    block_id = 6
    top_right = '1.5 1.5 1.8'
    bottom_left = '1.2 1.2 1.5'
  []
  [nodevug1]
    type = BoundingBoxNodeSet
    new_boundary = 'nodesvug1'
    bottom_left = '1.2 1.2 1.5'
    top_right = '1.5 1.5 1.8'
  []
  [vug2]
    type = SubdomainBoundingBox
    block_id = 7
    top_right = '8.5 8.5 8.8'
    bottom_left = '8.2 8.2 8.5'
  []
  [nodevug2]
    type = BoundingBoxNodeSet
    new_boundary = 'nodesvug2'
    bottom_left = '8.2 8.2 8.5'
    top_right = '8.5 8.5 8.8'
  []




*** ERROR ***
/home/p815l321/projects/sim/problems/newtest6.i:55.3: section '[MeshModifiers/vfrac1]' does not have an associated "Action".
 Common causes:
- you misspelled the Action/section name
- the app you are running does not support this Action/syntax
/home/p815l321/projects/sim/problems/newtest6.i:61.3: section '[MeshModifiers/nodevfrac1]' does not have an associated "Action".
 Common causes:
- you misspelled the Action/section name
- the app you are running does not support this Action/syntax
/home/p815l321/projects/sim/problems/newtest6.i:67.3: section '[MeshModifiers/vfrac2]' does not have an associated "Action".
 Common causes:
- you misspelled the Action/section name
- the app you are running does not support this Action/syntax
/home/p815l321/projects/sim/problems/newtest6.i:73.3: section '[MeshModifiers/nodevfrac2]' does not have an associated "Action".
 Common causes:
- you misspelled the Action/section name
- the app you are running does not support this Action/syntax
/home/p815l321/projects/sim/problems/newtest6.i:79.3: section '[MeshModifiers/vug1]' does not have an associated "Action".
 Common causes:
- you misspelled the Action/section name
- the app you are running does not support this Action/syntax

**Reply by GiudGiud** [reply]
MeshModifiers dont exist anymore.

Use meshgenerators instead
https://mooseframework.inl.gov/syntax/Mesh/index.html

**Comment by POLonge** [comment]
Hi @GiudGiud @1runer
What does this mean? The simulation was aborted at first time step

Time Step 0, time = 0

Postprocessor Values:
+----------------+----------------+----------------+
| time           | mineral_bh     | p_bh           |
+----------------+----------------+----------------+
|   0.000000e+00 |   0.000000e+00 |   0.000000e+00 |
+----------------+----------------+----------------+


Time Step 1, time = 1, dt = 1
 0 Nonlinear |R| = 8.193941e+12
munmap_chunk(): invalid pointer
Aborted (core dumped)

**Reply by GiudGiud** [reply]
that's a bug. You can run it in the debugger to find where
https://mooseframework.inl.gov/application_development/debugging.html

is this a recent version MOOSE? what does it say in the header of the simulation log?

**Comment by POLonge** [comment]
Hi @GiudGiud @1runer

I am trying to make an input to simulate flow in fractured vuggy rock. Let assume I have 2D or 3D mesh  how can I add a fractures/cavities or vugs into it.

**Reply by 1runer** [reply]
@POLonge 

Please see [THIS LINK](https://mooseframework.inl.gov/modules/porous_flow/nomultiapp_flow_through_fractured_media.html). It explains how to implement fractures as lower-dimensional elements (e.g. 2D planes in a 3D) and same-dimensional elements (3D thin fracture in matrix) and how you have to scale porosity and permeability according to the fracture aperture in order to be consistent with the fluid mass flowing therein. 

The corresponding example inputs for PorousFlow are located [HERE](https://github.com/idaholab/moose/tree/master/modules/porous_flow/examples/flow_through_fractured_media)

**Comment by POLonge** [comment]
Hi @GiudGiud @1runer

*** ERROR ***
The following error occurred in the object "Mesh", of type "SetupMeshAction".

Number of displacements must be greater than or equal to the dimension of the mesh!

What does this error mean?

**Reply by GiudGiud** [reply]
Hello

This means you set a displacements parameter but with less components than the dimension of the mesh
what dimension is your mesh again? Looks like 3D to me

**Reply by GiudGiud** [reply]
so displacements should be `displacements = 'disp_x disp_y disp_z'`

**Reply by POLonge** [reply]
Yes, it's 3D

**Reply by GiudGiud** [reply]
So is there any place the displacements are set to 2 variables instead?

and which object is throwing this error? This should be shown around the error message

**Reply by POLonge** [reply]
That did work.
Thanks for all the help and quick responses @GiudGiud @1runer.

Next steps. I want to write kernel based on darcy-brinksman model to simulates fluid flow in porous and free-flow region using single continuum. The free-flow which has mainly of vugs/cavity follows navier-stoke law and fractures while the porous media which consist of just matrix follows darcy's law.

I have the PDE's but with my low level C++ knowledge, I need assistance to go about writing: 
- various kernels, 
- userobject for BC at the interface between free flow and darcy flow region
-  adding them to moose.

**Reply by GiudGiud** [reply]
You ll definitely want to go over the moose tutorial once again and the application development tutorial to get more comfortable at writing kernels and C++. There s also a porous flow tutorial if you havent checked it out

**Comment by POLonge** [comment]
Hi,
@GiudGiud @1runer.

Please why this error?

Time Step 1, time = 100, dt = 100

    Computing Initial Residual
      Finished Computing Residual                                                        [  6.48 s] [ 1268 MB]
    Finished Computing Initial Residual                                                  [  6.48 s] [ 1268 MB]
    Finished Computing Residual                                                          [  6.38 s] [ 1268 MB]
 0 Nonlinear |R| = 1.841703e+03
    Computing Jacobian...free(): invalid size
Aborted (core dumped)

**Reply by GiudGiud** [reply]
are you running out of memory?

**Reply by POLonge** [reply]
Hi,
@GiudGiud @1runer.

Below is my computer spec.

Operating System: Kubuntu 22.04
KDE Plasma Version: 5.24.7
KDE Frameworks Version: 5.92.0
Qt Version: 5.15.3
Kernel Version: 5.15.0-91-generic (64-bit)
Graphics Platform: X11
Processors: 8 × Intel® Core™ i7-9700K CPU @ 3.60GHz
Memory: 15.4 GiB of RAM
Graphics Processor: NV137

How can divide the computations on different CPUs (parallel computing), to avoid this error.

**Comment by POLonge** [comment]
Hi,
@GiudGiud @1runer.

Below is my computer spec.

Operating System: Kubuntu 22.04
KDE Plasma Version: 5.24.7
KDE Frameworks Version: 5.92.0
Qt Version: 5.15.3
Kernel Version: 5.15.0-91-generic (64-bit)
Graphics Platform: X11
Processors: 8 × Intel® Core™ i7-9700K CPU @ 3.60GHz
Memory: 15.4 GiB of RAM
Graphics Processor: NV137

How can divide the computations on different CPUs (parallel computing), to avoid this error.

**Reply by GiudGiud** [reply]
are you using 1 process or multiple? If you are using `mpirun`, you need to use less ranks to consume less memory

**Reply by POLonge** [reply]
Could elaborate more. I didn't this

**Reply by GiudGiud** [reply]
ok how big is your mesh?
How many degrees of freedom does it report needing in the header of the simulation?
